iati/core/rulesets.py

```diff
@@ -386,36 +518,45 @@ def is_valid_for(self, dataset):
 
         Raises:
             AttributeError: When an argument is given that does not have the required attributes.
-            XPathEvalError(lxml.etree.XPathEvalError): When no valid XPath is available.
 
         """
-        compliant_paths = set()
+        context_elements = self._find_context_elements(dataset)
+        paths = set(self.paths)
+        compliant_paths = list()
+        no_of_paths = 0
 
-        for path in self.paths:
-            if len(dataset.xml_tree.xpath(path)) <= 1:
-                compliant_paths.add(path)
+        for context_element in context_elements:
+            if self._evaluate_condition(context_element):
+                return None
+            no_of_paths += len(paths)
+            for path in paths:
+                results = context_element.xpath(path)
+                if len(results) <= 1:
```

Comment:
Should there be an `else` with `return False`?

Reply:
The algorithm as a whole seems unnecessarily complex.

Reply:
There's no `return False` in `RuleNoMoreThanOne.is_valid_for`

Reply:
as for the algorithm, that is a problem for another branch me thinks